nengo/builder/optimizer.py

```diff
@@ -0,0 +1,969 @@
+from nengo.builder.neurons import SimNeurons
+from nengo.builder import operator
+from nengo.builder.operator import DotInc, ElementwiseInc, SlicedCopy
+from nengo.builder.signal import Signal
+from nengo.config import SupportRcDefaultsMixin
+from nengo.params import BoolParam, IntParam, RcDefault
+from nengo.utils.compat import zip_longest
+from nengo.utils.graphs import toposort, transitive_closure
+
+logger = logging.getLogger(__name__)
+
+
+class OpMergeOptimizer(SupportRcDefaultsMixin):
+    """Optimizes the operator graph by merging operators.
+
+    This reduces the number of iterators to iterate over in slow Python code
+    (as opposed to fast C code) and due to the merging operations will be
+    performed on larger chunks of sequential memory making better use of CPU
+    caching and prefetching.
+
+    The optimization algorithm has a worst case complexity of O(n^2 + e) where
+    n is the number of operators and e the number of edges in the dependency
+    graph. In practice the run time will be much better because not all n^2
+    pairwise combinations of operators will be evaluated. A grouping depending
+    on the operator type and view bases will be done with dictionaries. This
+    grouping can be done in amortized linear time and reduces the actual worst
+    case runtime of the optimization algorithm to O(g*m^2 + e) where g is the
+    number of groups and m the number of elements in a group. Moreover,
+    information about memory alignment will be used to cut the inner loop short
+    in many cases and gives a runtime much closer to linear in most cases.
+
+    Note that both `model` and `dg` will be changed by the `optimize` method.
+
+    Parameters
+    ----------
+    model : :class:`nengo.builder.builder.Model`
+        Builder output to optimize.
+    dg : dict
+        Dict of the form {a: {b, c}} where b and c depend on a specifying the
+        operator dependency graph of the model.
+    """
+
+    op_mergers = {}
+
+    enabled = BoolParam('enabled', default=True, optional=False)
+    max_passes = IntParam('max_passes', default=None, optional=True)
+
+    def __init__(self, model, dg):
+        self.enabled = RcDefault
+        self.max_passes = RcDefault
+
+        self.model = model
+        self.dg = dg
+
+        # Internal dict mapping Signals to Operators accessing those signals.
+        self._sig2op = None
+        # Internal set of Operators merged during the current optimization
+        # pass.
+        self._merged = None
+
+    @classmethod
+    def supports_merge(cls, op_type):
+        return op_type in cls.op_mergers
+
+    @classmethod
+    def can_merge_ops(cls, op1, op2):
+        return (
+            type(op1) == type(op2) and
+            type(op1) in cls.op_mergers and
+            cls.op_mergers[type(op1)].can_merge(op1, op2))
+
+    @classmethod
+    def merge_ops(cls, ops):
+        return cls.op_mergers[type(ops[0])].merge(ops)
+
+    @classmethod
+    def register_merger(cls, op_type):
```

Comment:
What is this decorator supposed to stop or enable? I'm getting a bit sleepy so it might be really obvious to me tomorrow morning.